tensorflow/compiler/xla/service/gpu/done_event_thunk.cc

```diff
@@ -0,0 +1,58 @@
+#include "tensorflow/compiler/xla/service/gpu/done_event_thunk.h"
+
+namespace xla {
+namespace gpu {
+
+using SpmdDoneEvents = absl::flat_hash_map<int, std::unique_ptr<se::Event>>;
+absl::flat_hash_map<int, SpmdDoneEvents*> all_done_events;
+
+void ResetEvents(size_t num_output) { all_done_events.clear(); }
+
+StatusOr<std::vector<std::unique_ptr<se::Event>>> TakeDoneEvents(
+    size_t output_index, int devices) {
+  auto it = all_done_events.find(output_index);
+  TF_RET_CHECK(it != all_done_events.end()) << "output index not found";
+  SpmdDoneEvents* output_done_events = it->second;
+  std::vector<std::unique_ptr<se::Event>> ret;
+  for (int i = 0;i < devices;++i) {
+    auto iter = output_done_events->find(i);
+    TF_RET_CHECK(iter != output_done_events->end()) << "done event not found";
+    ret.emplace_back(std::move(iter->second));
+    output_done_events->erase(iter);
+  }
+  all_done_events.erase(it);
```

Comment:
I am not sure whether one output uuid could correspond to multiple input uuids. If there is such case, we might be not allowed to erase here.

Reply:
let me explain the whole process by an example:
Mesh a:
1. Launch executable 'stage_1' with some outputs, each output corresponds to multiple `DeviceArray`s and exactly one uuid.
2. During the launch, each output creates a corresponding vector of `se::Event`s. So if you get events with index (i, j), you get the event for output `i` on local device `j`.
3. You take all events to your function and they are erased in `all_done_events`, then you unwrap `se::Event` to `CUevent` etc. and use them to launch sync and send kernel to communicate with Mesh b.
4. Launch executable `stage_1_bwd`. The `all_done_events` is overwritten here.
Mesh b:
launch a recv kernel for each tensor sent from Mesh a.

I'm not sure why you think the not allowed to erase happens. You can always manage them urself in step 3.

Reply:
What I mean about "one output uuid could correspond to multiple input uuids" is that: there are three executables A/B/C in one gpu. A outputs data_1, B inputs data_1 and C also inputs data_1.

Reply:
I may have found a case as bellow, do you think that makes sense? Or is it just a corner case that could be bypassed? 
When one tensor is calculated(A outputs data_1), it could both be sent to another host for resharding(B inputs data_1) and be freed(C also inputs data_1).

Reply:
Do you mean `all_done_events` should be clear after finishing each executable? We need save these cuda events to somewhere else. 
I thought it is a global variable and will keep the events after finishing one executable.

Reply:
Then you can take them after each launching executable and store them in your own class. It's dirty to do so in XLA runtime cuz we cannot get a hash key of (stage, micro-batch) pair in the runtime.

Reply:
Yes, that is a good implementation.